src/ucs/debug/instrument.h

```diff
@@ -88,28 +109,54 @@ static inline unsigned long ucs_instrument_expand_location(uint32_t location,
  */
 extern ucs_instrument_context_t ucs_instr_ctx;
 
+/*
+ * Register an instrumentation record - should be called once per record
+ * mentioned in the code, before the first record of each such mention is made.
+ *
+ * @param type          Instrumentation record type (to check if enabled).
+ * @param location_name Instrumentation record string (stored once).
+ *
+ * @return 0 for disabled record, positive instrumentation record id otherwise.
+ */
+uint32_t ucs_instrument_register(ucs_instrumentation_types_t type,
+                                 char const *location_name);
 
-void __ucs_instrument_record(uint64_t location, uint64_t lparam, uint32_t wparam);
+/*
+ * Store a new record with the given data.
+ *
+ * @param location Location id (provided by @ref ucs_instrument_register .
+ * @param lparam   64-bit user-defined data.
+ * @param wparam   32-bit user-defined data.
+ */
+void ucs_instrument_record(uint32_t location, uint64_t lparam, uint32_t wparam);
 
 
 /*
  * Helper macros
  */
-#define UCS_MAKE_LABEL(uniq) \
-    label_##uniq
-#define UCS_INSTRUMENT_RECORD_(uniq, lparam, wparam, ...) \
+#define UCS_INSTRUMENT_RECORD_(_type, _name, _lparam, _wparam, ...) \
     do { \
-        if (ucs_instr_ctx.enable) do { \
-    UCS_MAKE_LABEL(uniq):\
-            __ucs_instrument_record((uint64_t)&&UCS_MAKE_LABEL(uniq), (uint64_t)(lparam), (uint32_t)(wparam)); \
-        } while (0); \
+        static uint32_t instrument_id = (uint32_t)-1; \
+        if (instrument_id) { \
+            if (ucs_unlikely(instrument_id == (uint32_t)-1)) { \
+                char instr_name[LOCATION_NAME_LEN] = {0}; \
```

Comment:
IMHO it would be better to pass __FILE__ and __LINE__ as arguments to ucs_instrument_register instead of constructing the location string in the macro itself.